From 067de7270b3433afc244ba4a607f3491f0d0fc5c Mon Sep 17 00:00:00 2001
From: Nicolas Ferre <nicolas.ferre@atmel.com>
Date: Fri, 16 Nov 2012 16:47:44 +0100
Subject: net/macb: align ring buffer function with mainline

Available number of descriptors in ring buffer function by David Laight.

Signed-off-by: Nicolas Ferre <nicolas.ferre@atmel.com>
---
 drivers/net/ethernet/cadence/macb.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/cadence/macb.c b/drivers/net/ethernet/cadence/macb.c
index e3168bf..1f7b3a3 100644
--- a/drivers/net/ethernet/cadence/macb.c
+++ b/drivers/net/ethernet/cadence/macb.c
@@ -31,12 +31,12 @@
 #include "macb.h"
 
 #define RX_BUFFER_SIZE		128
-#define RX_RING_SIZE		512
+#define RX_RING_SIZE		512 /* must be power of 2 */
 #define RX_RING_BYTES		(sizeof(struct macb_dma_desc) * RX_RING_SIZE)
 #define RX_BUFFERS_PER_PAGE	(PAGE_SIZE / RX_BUFFER_SIZE)
 #define RX_RING_PAGES		(RX_RING_SIZE / RX_BUFFERS_PER_PAGE)
 
-#define TX_RING_SIZE		128
+#define TX_RING_SIZE		128 /* must be power of 2 */
 #define TX_RING_BYTES		(sizeof(struct macb_dma_desc) * TX_RING_SIZE)
 
 /* minimum number of free TX descriptors before waking up TX process */
@@ -63,7 +63,7 @@ static unsigned int macb_tx_ring_wrap(unsigned int index)
 
 static unsigned int macb_tx_ring_avail(struct macb *bp)
 {
-	return TX_RING_SIZE - (bp->tx_head - bp->tx_tail);
+	return (bp->tx_tail - bp->tx_head) & (TX_RING_SIZE - 1);
 }
 
 static struct macb_dma_desc *macb_tx_desc(struct macb *bp, unsigned int index)
-- 
1.8.0.197.g5a90748

